lib/sass_builder.dart

```diff
@@ -1,94 +1,148 @@
+    var cssOutput = compile(tempAssetPath,
+        packageResolver: new SyncPackageResolver.root(tempDir.packagesDir.uri));
+
+    // Write the builder output
+    var outputId = inputId.changeExtension(_outputExtension);
+    buildStep.writeAsString(outputId, '${cssOutput}\n');
+    _log.fine('wrote css file: ${outputId.path}');
   }
 
   @override
-  Future build(BuildStep buildStep) async {
-    var inputId = buildStep.inputId;
-    _log.info('processing changed file: ${inputId}');
+  Map<String, List<String>> get buildExtensions => {
+        '.scss': [_outputExtension],
+        '.sass': [_outputExtension],
+      };
 
-    if (!basename(inputId.path).startsWith('_')) {
-      if (!mainInputs.contains(inputId)) {
-        mainInputs.add(inputId);
-        _saveMainInputs();
-      }
+  // Reads `id` and all transitive imports while copying them to `tempDir`.
+  //
+  // Uses `buildStep to read the assets and the `ScratchSpace` API to write
+  // `tempDir`.
+  Future _readAndCopyImports(
+      AssetId id, BuildStep buildStep, ScratchSpace tempDir) async {
+    var assetsToCopy = new Queue<AssetId>();
+    assetsToCopy.add(id);
+
+    while (assetsToCopy.isNotEmpty) {
+      id = assetsToCopy.removeFirst();
+
+      var contents = await buildStep.readAsString(id);
+      _log.fine('read file: ${id}');
+      await tempDir.ensureAssets([id], buildStep);
+
+      var imports = await _importedAssets(id, contents, buildStep);
+      assetsToCopy.addAll(imports);
+      var importLog = imports.fold('', (acc, import) => '$acc\n  ${import}');
+      _log.fine('found imports:$importLog');
     }
-    _log.info('using mainInputs: $mainInputs');
-
-    if (mainInputs.isNotEmpty) {
-      var mainInputsToRemove = new Set();
-      for (var mainInput in mainInputs) {
-        _log.info('parsing: ${mainInput}');
-        if (!new File(mainInput.path).existsSync()) {
-          _log.info('adding mainInput to remove: $mainInput');
-          mainInputsToRemove.add(mainInput);
+  }
+
+  // Returns the `AssetId`s of all the scss imports in `contents`.
+  Future<Iterable<AssetId>> _importedAssets(
+      AssetId id, String contents, BuildStep buildStep) async {
+    var importedAssets = new Set<AssetId>();
+
+    for (var importBlock in _importBlockRegExp
+        .allMatches(contents.replaceAll(_sassCommentRegExp, ''))) {
+      var imports = _fileNameRegExp.allMatches(importBlock.group(1));
+      for (var import in imports) {
+        var importId = new AssetId(_importPackage(import.group(1), id.package),
+            _importPath(import.group(1), id.path));
+
+        if (!await (buildStep.canRead(importId))) {
+          // Try same asset path except filename starting with an underscore.
+          _log.fine('could not read file: ${importId}');
+          importId = new AssetId(importId.package,
+              join(dirname(importId.path), '_${basename(importId.path)}'));
+        }
+
+        if (!await buildStep.canRead(importId)) {
```

Comment:
I would move this logic inside the if above, otherwise you are often checking the same import twice.